src/Abc.Zebus/Directory/PeerDirectoryClient.cs

```diff
@@ -294,56 +292,52 @@ public void Handle(PeerStarted message)
+            var snapshotGeneratingMessageTypes = _messageDispatcher.GetMessageHanlerInvokers()
+                                                                   .Select(x => x.MessageHandlerType.GetBaseTypes().SingleOrDefault(y => y.IsGenericType && y.GetGenericTypeDefinition() == typeof(SubscriptionHandler<>))?.GenericTypeArguments[0])
+                                                                   .Where(x => x != null)
+                                                                   .ToHashSet();
+
+            Thread.MemoryBarrier();
+            _snapshotGeneratingMessageTypes = snapshotGeneratingMessageTypes;
         }
 
         private void DispatchSubscriptionUpdatedMessages(PeerId peerId, Subscription[] subscriptions)
         {
             if (peerId == _self.Id)
                 return;
 
-            lock (_snapshotGeneratingMessageTypes)
+            var messageContext = GetMessageContextForSubscriptionUpdated();
+
+            foreach (var subscription in subscriptions)
             {
-                var messageContext = GetMessageContextForSubscriptionUpdated();
-                foreach (var subscription in subscriptions)
-                {
-                    if (!_snapshotGeneratingMessageTypes.Contains(subscription.MessageTypeId.GetMessageType()))
-                        continue;
+                if (!_snapshotGeneratingMessageTypes.Contains(subscription.MessageTypeId.GetMessageType()))
+                    continue;
 
-                    var subscriptionUpdatedMessage = new SubscriptionUpdatedMessage(subscription, peerId);
-                    _messageDispatcher.Dispatch(new MessageDispatch(messageContext, subscriptionUpdatedMessage, null, (dispatch, result) => { }));
-                }
+                var subscriptionUpdatedMessage = new SubscriptionUpdatedMessage(subscription, peerId);
+                _messageDispatcher.Dispatch(new MessageDispatch(messageContext, subscriptionUpdatedMessage, null, (dispatch, result) => { }));
             }
         }
 
-        private MessageContext GetMessageContextForSubscriptionUpdated() => MessageContext.Current ?? MessageContext.CreateOverride(_self.Id, _self.EndPoint);
-
         private void DispatchSubscriptionUpdatedMessages(PeerId peerId, SubscriptionsForType[] subscriptions)
         {
             if (peerId == _self.Id)
                 return;
 
-            lock (_snapshotGeneratingMessageTypes)
+            var messageContext = GetMessageContextForSubscriptionUpdated();
+
+            foreach (var subscription in subscriptions)
             {
-                var messageContext = GetMessageContextForSubscriptionUpdated();
-                foreach (var subscription in subscriptions)
-                {
-                    if (!_snapshotGeneratingMessageTypes.Contains(subscription.MessageTypeId.GetMessageType()))
-                        continue;
+                if (!_snapshotGeneratingMessageTypes.Contains(subscription.MessageTypeId.GetMessageType()))
```

Comment:
This will give incoherent results as the reference `_snapshotGeneratingMessageTypes` can change during the iteration.